File: 3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        # Pair elements with their original indices and sort by value
        sorted_nums = sorted([(num, i) for i, num in enumerate(nums)])
        
        n = len(nums)
        ans = [0] * n
        i = 0
        
        while i < n:
            # Find the boundary of the current connected component
            j = i + 1
            while j < n and sorted_nums[j][0] - sorted_nums[j-1][0] <= limit:
                j += 1
            
            # Extract the current component
            component = sorted_nums[i:j]
            
            # Extract and sort the original indices of this component
            indices = sorted([idx for val, idx in component])
            
            # Place the sorted values into the sorted original indices
            for k in range(len(component)):
                ans[indices[k]] = component[k][0]
                
            # Move to the start of the next component
            i = j
            
        return ans

File: 3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.cpp
#include <vector>
#include <algorithm>
#include <utility>

using namespace std;

class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        int n = nums.size();
        
        // Pair up each number with its original index
        vector<pair<int, int>> sorted_nums(n);
        for (int i = 0; i < n; ++i) {
            sorted_nums[i] = {nums[i], i};
        }
        
        // Sort the pairs based on the values
        sort(sorted_nums.begin(), sorted_nums.end());
        
        vector<int> result(n);
        int i = 0;
        
        // Group elements into connected components
        while (i < n) {
            int j = i + 1;
            // Extend the component as long as the difference between adjacent sorted elements is <= limit
            while (j < n && sorted_nums[j].first - sorted_nums[j - 1].first <= limit) {
                j++;
            }
            
            // Extract all original indices for the current component
            vector<int> indices;
            for (int k = i; k < j; ++k) {
                indices.push_back(sorted_nums[k].second);
            }
            
            // Sort the indices so we can place the smallest values into the earliest available slots
            sort(indices.begin(), indices.end());
            
            // Re-assign the sorted values to the sorted indices in the result array
            for (int k = 0; k < indices.size(); ++k) {
                result[indices[k]] = sorted_nums[i + k].first;
            }
            
            // Move to the start of the next component
            i = j;
        }
        
        return result;
    }
};